src/bpfmap.cpp

```diff
@@ -60,17 +54,155 @@ bool BpfMap::is_per_cpu_type() const
          type() == libbpf::BPF_MAP_TYPE_LRU_PERCPU_HASH;
 }
 
-bool BpfMap::is_clearable() const
-{
-  return is_bpf_map_clearable(type());
-}
-
 bool BpfMap::is_printable() const
 {
   // Internal maps are not printable
   return bpf_name().starts_with("AT_");
 }
 
+KeyVec BpfMap::collect_keys() const
+{
+  uint8_t *old_key = nullptr;
+  auto key = KeyType(key_size_);
+
+  // snapshot keys, then operate on them
+  KeyVec keys;
+  while (bpf_map_get_next_key(fd(), old_key, key.data()) == 0) {
+    keys.push_back(key);
+    old_key = key.data();
+  }
+  return keys;
+}
+
+Result<OK> BpfMap::zero_out(int nvalues) const
+{
+  auto keys = collect_keys();
+  auto value_size = static_cast<size_t>(value_size_) *
+                    static_cast<size_t>(nvalues);
+  ValueType zero(value_size, 0);
+  for (auto &k : keys) {
+    int err = bpf_map_update_elem(fd(), k.data(), zero.data(), BPF_EXIST);
+
+    if (err && err != -ENOENT) {
+      return make_error<BpfMapError>(
+          "failed to zero map: " + std::to_string(err) + " (" +
+          std::strerror(-err) + ")");
+    }
+  }
+  return OK();
+}
+
+Result<OK> BpfMap::clear(int nvalues) const
+{
+  if (!is_bpf_map_clearable(type())) {
+    return zero_out(nvalues);
+  }
+  auto keys = collect_keys();
+  for (auto &k : keys) {
+    int err = bpf_map_delete_elem(fd(), k.data());
+    if (err && err != -ENOENT) {
+      return make_error<BpfMapError>(
+          "failed to clear map: " + std::to_string(err) + " (" +
+          std::strerror(-err) + ")");
+    }
+  }
+  return OK();
+}
+
+Result<> BpfMap::update_elem(const void *key, const void *value) const
+{
+  auto err = bpf_map_update_elem(fd(), key, value, BPF_ANY);
+  if (err != 0) {
+    return make_error<BpfMapError>(
+        "failed to update elem: " + std::to_string(err) + " (" +
+        std::strerror(-err) + ")");
+  }
+  return OK();
+}
+
+Result<> BpfMap::lookup_elem(const void *key, void *value) const
+{
+  auto err = bpf_map_lookup_elem(fd(), key, value);
+  if (err != 0) {
+    return make_error<BpfMapError>(
+        "failed to lookup elem: " + std::to_string(err) + " (" +
+        std::strerror(-err) + ")");
+  }
+  return OK();
+}
+
+Result<MapElements> BpfMap::collect_elements(int nvalues) const
+{
+  uint8_t *old_key = nullptr;
+  auto key = KeyType(key_size_);
+  MapElements values_by_key;
+
+  while (bpf_map_get_next_key(fd(), old_key, key.data()) == 0) {
+    auto value = ValueType(static_cast<size_t>(value_size_) *
+                           static_cast<size_t>(nvalues));
+    int err = bpf_map_lookup_elem(fd(), key.data(), value.data());
+    if (err == -ENOENT) {
+      // key was removed by the eBPF program during bpf_map_get_next_key() and
+      // bpf_map_lookup_elem(), let's skip this key
+      continue;
+    } else if (err) {
+      return make_error<BpfMapError>(
+          "failed to look up elem: " + std::to_string(err) + " (" +
+          std::strerror(-err) + ")");
```

Comment:
Just for context (feel free to do this or not) for future errors, you could just put the structure in the info class and format once in the log() function.

For example, the class could have "std::string op", and "int err" and you could construct `make_error<BpfMapError>("lookup", -err)`. It's just pushing the structure into the types. Makes it so there's less boilerplate formatting code around, etc.